### PR TITLE
Add field index options and query dsl for highlight

### PR DESCRIPTION
**Pull request by CCheSumo**
- Added capability to specify index_options in field mapping 
  https://www.elastic.co/guide/en/elasticsearch/reference/current/index-options.html
- Added query dsl for highlight
  https://www.elastic.co/guide/en/elasticsearch/reference/1.5/search-request-highlighting.html#postings-highlighter
- Added highlight results in SearchResponse

### CODE REVIEW

**Comment by CCheSumo** [comment]
@rcoh Can you take a look at this code changes? It enables "highlight" in the query.

**Comment by rcoh** [comment]
Will take a look this evening.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala:305]: please file an issue to remove this default in 2.0

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala]: This can just go inside `MappingDsl` and then it can be sealed

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala]: why not call this `indexOption` to be consistent with the other terms?

File: elasticsearch-core/src/test/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClientTest.scala
Comment (rcoh) [inline elasticsearch-core/src/test/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClientTest.scala:107]: This test only checks that you put the mapping and it happens to not error (if we even propagate errors properly!) Can you make a call to `getMapping` to verify that the mapping was actually applied?

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala:378]: maybe I'm missing something -- can we just have HighlightRoot extend RootObject rather than creating a separate `QueryRootBase`?

Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala]: maybe I'm missing something -- can we just have `HighlightRoot` extend RootObject rather than creating a separate `QueryRootBase`?

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala]: this code is pretty hard to read. I would probably do it as:
`fields.map(_.toJson).reduce(_ ++ _)` what do you think?

**Review by rcoh — COMMENTED** [review]
Need to take a bit more time to read all of it. Some comments so far.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClient.scala:305]: Issue filed as requested https://github.com/SumoLogic/elasticsearch-client/issues/77

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala]: Moved to be a sealed trait in MappingDsl as requested.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala]: Updated as requested.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala:378]: Oh, making a separate `QueryRootBase` is to ensure `query` must take one of its two implementations (either `QueryRoot`, or `HighlightRoot`) and throw error at compile time if other type of parameter is provided.

Comment (CCheSumo) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala]: Not sure why this is duplicated. Please refer to the comments above.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala]: Ah, I should have thought of using reduce instead of foldLeft. It makes the code much cleaner! Thanks for the advice. Updated.

File: elasticsearch-core/src/test/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClientTest.scala
Comment (CCheSumo) [reply elasticsearch-core/src/test/scala/com/sumologic/elasticsearch/restlastic/RestlasticSearchClientTest.scala:107]: Added code to verify mapping using getMapping.

**Comment by CCheSumo** [comment]
@rcoh Thanks for the feedbacks. I should have addressed most of the concerns raised. I will wait for your second review before merging.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala]: seems like this could move into QueryDsl as well since it's only used there?

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala
Comment (rcoh) [inline elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/MappingDsl.scala:100]: I think we need a formalized way of handling elasticsearch 2.x stuff. Ideally it would be a different DSL you could import or something. Another task to think about pre-2.0.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (rcoh) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala:378]: I see -- you don't want a HighlightRoot to be nested inside of a HighlightRoot

Comment (rcoh) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala:378]: Can you rename it to `QueryRootLike`? That's the scala standard way for saying something is a trait.

**Review by rcoh — APPROVED** [review]
LGTM -- see a couple more small comments above.

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/DslCommons.scala]: Moved to QueryDsl as a sealed trait

File: elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala
Comment (CCheSumo) [reply elasticsearch-core/src/main/scala/com/sumologic/elasticsearch/restlastic/dsl/QueryDsl.scala:378]: Updated QueryRootBase to QueryRootLike

**Pull request merged**
